### PR TITLE
Support for animated formats (gif/webp/apng)

### PR DESCRIPTION
**Pull request by benzkji**
A first draft. Test still passing, no new tests for the moment.

I don't think that a wrapper (processors leaving as is, abstract things away for processors) would be a good solution, as some processors could still not operate like this. For example, the autocrop would autocrop each frame individually, and if a processor creates a new Image(), and pastes the original into it, frames would still be lost - as the `background` processor does it. 

With my proposal, animated GIF support would need to be supported by each processor individually. Most of the times, this would mean a simple usage of the new helper function instead of directly calling `im.whatever()`.

Using 
```
THUMBNAIL_IMAGE_SAVE_OPTIONS = {
    "GIF": {"save_all": True}, 
}
THUMBNAIL_PRESERVE_EXTENSIONS = ("gif", )
```
And animated GIFs work (on my machine), hopefully animated WEBP would as well. A first feedback for the approach is very welcome.

### CODE REVIEW

File: easy_thumbnails/processors.py [hunk 1]
```diff
@@ -35,6 +35,27 @@ def _points_table():
             yield j
 
 
+def _call_pil_method(im, method, *args, **kwargs):
```
Comment (SmileyChris) [inline easy_thumbnails/processors.py:38]: I'd suggest a wrapper class that looked something like:
```py
from functools import partial

class FrameAware:
    def __new__(cls, im):
        if getattr(im, "n_frames", 1) > 1:
            return super().__new__(cls)
        return im

    def __init__(self, im):
        self.im = im

    def apply_to_frames(self, method, *args, **kwargs):
        results = []
        for i in range(self.im.n_frames):
            self.im.seek(i)
            result = method(*args, **kwargs)
            self.im.paste(result)
        return self.im

    def __getattr__(self, key):
        method = getattr(self.im, key)
        return partial(self.apply_to_frames, method)
```
Now you can do:
```py
im = FrameAware(im).convert(new_mode)
```

Comment (benzkji) [reply easy_thumbnails/processors.py:38]: Ah, looks nice. Will try it.

**Comment by benzkji** [comment]
...also, will remove _call_pil_method soon.

**Comment by benzkji** [comment]
It has tests now, for all builtin processors. Though, it only tests if frames are preserved, and some of the basic functionalities (as resize/crop) -  I'm afraid I don't fully understand all of the processors, and especially not all of the edge cases involved.

If the used BytesIO workaround is ok for @SmileyChris I think it could be merged?

**Comment by benzkji** [comment]
...remaining to add: docs. Also, I would add `save_all=True` for all images with more than one frame, it would be better to support them without having to set/use `THUMBNAIL_IMAGE_SAVE_OPTIONS`.

**Comment by benzkji** [comment]
...also, docs: how to make them work again?

**Comment by benzkji** [comment]
@SmileyChris feedback welcome :) or maybe shall I reach out to @jrief ?

**Comment by jrief** [comment]
@benzkji Hi Ben, you're changing code and code-styling in the same pull request. This makes a review extremly difficult to understand. Can this be split into separate PRs?

And btw., I prefer single quotes for strings which are not read by humans, but this is my personal opinion.

**Comment by benzkji** [comment]
@jrief Hi Jacob. Sorry, that slipped in, I probably had black active on save. Will try to revert these changes.

**Comment by benzkji** [comment]
things remaining:

- wether to enable "save_all" by default if multiple frames are detected?
- write docs
- test in real live (done that a few months before with django-filer, but it did not always work...), with webp, etc.

**Comment by benzkji** [comment]
real world feedback with django-filer:

- animated png/gif/webp all work, when preserving their extensions and setting save options
- some animated gifs could not be saved as WEBP, because of a problem during saving the image in PIL.
- but in general, gif>web and png>webp was working
- bigger images could block your server

**Comment by benzkji** [comment]
Thinking about it, it would be nice if one could define if animations should be preserved, when using easy-thumbnails...not an "always on/off" solution, but rather some kind of `crop=1 animated=1/0` ?! @jrief next step?!

**Comment by benzkji** [comment]
if anyone interested:

`PIL/WebPImagePlugin.py, line 199, in _save_all` > `not enough values to unpack (expected 3, got 0)` is the error, when trying a special GIF (probably because of a reduced palette) to .webp.

**Comment by benzkji** [comment]
but, without "save_all" enabled for everyone automatically, I guess this could be released as a beta, if no issues arise. People could check manually, with preserve extensions and save_all, if it works. hopefully no existing features would be broken, aka we have a testsuite - I don't know how extensive it is?!

File: easy_thumbnails/processors.py [hunk 2]
```diff
@@ -108,7 +149,8 @@ def autocrop(im, autocrop=False, **kwargs):
         bg = Image.new('L', im.size, 255)
         bbox = ImageChops.difference(bw, bg).getbbox()
         if bbox:
-            im = im.crop(bbox)
+            # im = im.crop(bbox)
```
Comment (jrief) [inline easy_thumbnails/processors.py:152]: please remove commented code

File: easy_thumbnails/processors.py [hunk 3]
```diff
@@ -274,7 +318,9 @@ def scale_and_crop(im, size, crop=False, upscale=False, zoom=None, target=None,
                     diff_y = diff_y - add - remove
                 box = (left, top, right, bottom)
             # Finally, crop the image!
-            im = im.crop(box)
+            # im = im.crop(box)
```
Comment (jrief) [inline easy_thumbnails/processors.py:321]: detto

File: easy_thumbnails/processors.py [hunk 4]
```diff
@@ -291,9 +337,11 @@ def filters(im, detail=False, sharpen=False, **kwargs):
 
     """
     if detail:
-        im = im.filter(ImageFilter.DETAIL)
+        # im = im.filter(ImageFilter.DETAIL)
```
Comment (jrief) [inline easy_thumbnails/processors.py:340]: detto

**Review by jrief — CHANGES_REQUESTED** [review]
Thanks for this improvement.

Would be great if you could add a few lines of documentation and mention this feature in the ChangeLog.

**Comment by benzkji** [comment]
@jrief done so (docs & changelog). BTW, do you know what must be done to get readthedocs up and running again? https://easy-thumbnails.readthedocs.io/en/latest/ is for 2.7, no word of SVG et al?

**Comment by benzkji** [comment]
ah, shall I merge master into my branch? My current local tests are passing (without python3.7 ;), but that's not up to date at all.

**Comment by benzkji** [comment]
thx. will check the testsuite.

**Comment by jrief** [comment]
@benzkji I assume that the local behavior of colorspace is different from what the tests on GitHub-Actions tell us. Any idea what could be the problem?

**Comment by benzkji** [comment]
not yet. After the merge it fails localy on my machine as well.

**Comment by benzkji** [comment]
test pass now. it is strange, though, it seems PIL eliminates empty frames. but also, my test PNG and WEBPs seem empty (when saving them out), but when using filer it works, and in tests it works as well, now. Looks like a "writing the tests is the hardest part" problem.
